packages/firestore/src/api/reference_impl.ts

```diff
@@ -1095,87 +1095,140 @@ function onSnapshotBundle<AppModelType, DbModelType extends DocumentData>(
+      };
     }
-  } else if (json.bundleSource === 'DocumentSnapshot') {
+    return onSnapshotQuerySnapshotBundle(
+      db,
+      snapshotJson,
+      options,
+      observer!,
+      args[curArg] as FirestoreDataConverter<DbModelType>
+    );
+  } else if (snapshotJson.bundleSource === 'DocumentSnapshot') {
+    let observer: {
+      next: (snapshot: DocumentSnapshot<AppModelType, DbModelType>) => void;
+      error?: (error: FirestoreError) => void;
+      complete?: () => void;
+    } | null = null;
     if (typeof args[curArg] === 'object' && isPartialObserver(args[1])) {
       const userObserver = args[curArg++] as PartialObserver<
         DocumentSnapshot<AppModelType, DbModelType>
       >;
-      return onSnapshotDocumentSnapshotBundle(
-        db,
-        json,
-        options,
-        {
-          next: userObserver.next!,
-          error: userObserver.error,
-          complete: userObserver.complete
-        },
-        args[curArg] as FirestoreDataConverter<DbModelType>
-      );
+      observer = {
+        next: userObserver.next!,
+        error: userObserver.error,
+        complete: userObserver.complete
+      };
     } else {
-      return onSnapshotDocumentSnapshotBundle(
-        db,
-        json,
-        options,
-        {
-          next: args[curArg++] as (
-            snapshot: DocumentSnapshot<AppModelType, DbModelType>
-          ) => void,
-          error: args[curArg++] as (error: FirestoreError) => void,
-          complete: args[curArg++] as () => void
-        },
-        args[curArg] as FirestoreDataConverter<DbModelType>
-      );
+      observer = {
+        next: args[curArg++] as (
+          snapshot: DocumentSnapshot<AppModelType, DbModelType>
+        ) => void,
+        error: args[curArg++] as (error: FirestoreError) => void,
+        complete: args[curArg++] as () => void
+      };
     }
+    return onSnapshotDocumentSnapshotBundle(
+      db,
+      snapshotJson,
+      options,
+      observer!,
+      args[curArg] as FirestoreDataConverter<DbModelType>
+    );
   } else {
     throw new FirestoreError(
       Code.INVALID_ARGUMENT,
-      `unsupported bundle source: ${json.bundleSource}`
+      `unsupported bundle source: ${snapshotJson.bundleSource}`
     );
   }
 }
 
+/**
+ * Ensures the data required to construct an {@link onSnapshot} listener exist in a `snapshotJson`
+ * object that originates from {@link DocumentSnapshot.toJSON} or {@link Querysnapshot.toJSON}. The
+ * data is normalized into a typed object.
+ *
+ * @param snapshotJson - The JSON object that the app provided to {@link onSnapshot}.
+ * @returns A normalized object that contains all of the required bundle JSON fields. If
+ * {@link snapshotJson} doesn't contain the required fields, or if the fields exist as empty
+ * strings, then the {@link snapshotJson.error} field will be a non empty string.
+ *
+ * @internal
+ */
+function normalizeSnapshotJsonFields(snapshotJson: object): {
```

Comment:
I might have used a [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) for this. However, since you are also checking values, this implementation is also okay, IMO.